rpc/server.go

```diff
@@ -51,25 +61,51 @@ type Server struct {
 	addr           net.Addr              // Server address; may change if picking unused port
 	closed         bool                  // Set upon invocation of Close()
 	closeCallbacks []func(conn net.Conn) // Slice of callbacks to invoke on conn close
+	methods        map[string]method
 }
 
 // NewServer creates a new instance of Server.
 func NewServer(addr net.Addr, context *Context) *Server {
 	s := &Server{
-		Server:  rpc.NewServer(),
 		context: context,
 		addr:    addr,
+		methods: map[string]method{},
 	}
 	heartbeat := &HeartbeatService{
 		clock:              context.localClock,
 		remoteClockMonitor: context.RemoteClocks,
 	}
-	if err := s.RegisterName("Heartbeat", heartbeat); err != nil {
+	if err := heartbeat.Register(s); err != nil {
 		log.Fatalf("unable to register heartbeat service with RPC server: %s", err)
 	}
 	return s
 }
 
+// Register a new method handler. `name` is a qualified name of the form "Service.Name".
+// `handler` is a function that takes an argument of the same type as `reqPrototype`.
+// Both the argument and return value of 'handler' should be a pointer to a protocol
+// message type.
+func (s *Server) Register(name string, handler func(interface{}) (interface{}, error),
+	reqPrototype interface{}) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if _, ok := s.methods[name]; ok {
+		return util.Errorf("method %s already registered", name)
+	}
+	reqType := reflect.TypeOf(reqPrototype)
+	if reqType.Kind() != reflect.Ptr {
```

Comment:
Changing `interface{}` to `proto.Message` in the handler signature results in import conflicts: any file that defines a handler needs to deal with our `proto` package; if we require `proto.Message` in handler signatures then they'll also need to import that `proto` package under an alias (`gogoproto`). This is pretty ugly, but probably worthwhile to avoid surprises with pointers versus non-pointers.